lib/tests-tree-builder/base.ts

```diff
@@ -241,26 +241,59 @@ export class BaseTestsTreeBuilder {
         }
 
         const suite = this._tree.suites.byId[suiteId];
+        const uniqueChildStatuses: TestStatus[] = [];
+
+        for (const browserId of suite.browserIds || []) {
+            if (!browserId) {
+                continue;
+            }
 
-        const resultStatuses = _.compact(([] as (string | undefined)[]).concat(suite.browserIds))
-            .map((browserId: string) => {
-                const browser = this._tree.browsers.byId[browserId];
-                const lastResultId = _.last(browser.resultIds) as string;
+            const browserResultIds = this._tree.browsers.byId[browserId].resultIds;
+            const lastResultId = browserResultIds[browserResultIds.length - 1];
+            const lastResultStatus = this._tree.results.byId[lastResultId].status;
 
-                return this._tree.results.byId[lastResultId].status;
-            });
+            if (lastResultStatus && !uniqueChildStatuses.includes(lastResultStatus)) {
+                uniqueChildStatuses.push(lastResultStatus);
+            }
+        }
+
+        for (const childSuiteId of suite.suiteIds || []) {
+            if (!childSuiteId) {
+                continue;
+            }
 
-        const childrenSuiteStatuses = _.compact(([] as (string | undefined)[]).concat(suite.suiteIds))
-            .map((childSuiteId: string) => this._tree.suites.byId[childSuiteId].status);
+            const childSuiteStatus = this._tree.suites.byId[childSuiteId].status;
+
+            if (childSuiteStatus && !uniqueChildStatuses.includes(childSuiteStatus)) {
+                uniqueChildStatuses.push(childSuiteStatus);
+            }
+        }
 
-        const status = determineFinalStatus(_.compact([...resultStatuses, ...childrenSuiteStatuses]));
+        // Updating parent suites status, implying "finalStatus(A, A, A, B) === finalStatus(A, B)"
+        suite.status = determineFinalStatus(uniqueChildStatuses) || void 0;
 
-        // if newly determined status is the same as current status, do nothing
-        if (suite.status === status) {
+        if (!suite.status) {
             return;
         }
 
-        suite.status = status || undefined;
-        this._setStatusForBranch(testPath.slice(0, -1));
```

Comment:
For each suite we would fully recalculate its parent state
But given that "determineFinalStatus(determineFinalStatus(A, B), C) === determineFinalStatus(A, B, C)", we can only recalculate parent status with "determineFinalStatus(parentStatus, newChildStatus)", because parentStatus is result of "determineFinalStatus" being called on all other nodes